### DISCUSSION TITLE
`populate_existing()` without overriding end-user changes on attributes

### DISCUSSION THREAD

**Discussion by ay-ramp** [sqlalchemy/sqlalchemy#11162] [Usage Questions]
Hi there, we maintain a large codebase that uses SQLAlchemy and recently ran into some unexpected issues related to populate_existing() with with_for_update(). We started considering mandating usage of populate_existing() in our codebase any time someone calls with_for_update(), but then got nervous about _other_ unexpected behavior there around potentially unexpectedly overwriting dirty instance state.

We saw [this comment](https://github.com/sqlalchemy/sqlalchemy/issues/5572#issuecomment-690365997) from @zzzeek suggesting what looks like our ideal behavior here -- is this still potentially planned? And is there any mechanism for us to possibly get "populate_existing() except don't overwrite dirty attributes" in the meantime?

**Comment by zzzeek** [comment]
the mechanics of populate existing happen in the core of the attribute loading logic and they are hardcoded right now to do all attributes.    

if you have not actually encountered the case of non-flushed state being overwritten you are probably fine, what you could do would be to make sure there's always a flush() ahead of any query that uses populate_existing().

**Reply by ay-ramp** [reply]
Makes sense -- we probably will try to do something like that

After noticing this, we have actually realized we have some other errors in our app that are related to _not_ doing populates_existing() logic with successive fetches not related to `.for_updates()` (can go into these if it's helpful / you think we're missing something). Are you still potentially planning on a "`populate_existing()` without end user attribute changes" by default happening?

**Reply by zzzeek** [reply]
it was only a hypothetical idea, if it were a thing that came up.  so far, it has never come up, amazingly.   it seems in your case it's a concern but also hasnt come up :)   so, seems risky to just change it.    making it some kind of option, maybe, but now that it's `execution_options(populate_existing=True)` mmmm, I guess maybe it could be `execution_options(populate_existing=True, preserve_pending_attrs=True)`

**Reply by CaselIT** [reply]
Wouldn't pending attrs only be a consern if autoflush is set to false? If it's left to the default of true running any kind of query would trigger a flush.

Am I missing something here?

**Reply by zzzeek** [reply]
right!  that is part of it, and I guess people keep autoflush at True.  so yes

**Reply by ay-ramp** [reply]
Ah makes a ton of sense re: auto-flush. Makes sense why this isn't an issue.

In testing, have run into another issue around `populate_existing` that would be good to get your take on (let me know if it'd be better to open up a new discussion, happy to):

```
from typing import Any

from sqlalchemy import Column, Integer, create_engine, select, event, literal, func
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import (
    Session,
    query_expression,
    with_expression, deferred, undefer, relationship,
)

Base: Any = declarative_base()

class A(Base):
    __tablename__ = "a"
    id = Column(Integer, primary_key=True)
    other_column = deferred(Column(Integer), raiseload=True)
    x = query_expression("undefined")

e = create_engine("sqlite://", echo=False)
Base.metadata.create_all(e)

session = Session(e)



def run(populate_existing=False):
    a = A(id=1, other_column=5)
    session.add(a)
    session.commit()

    # print(a.other_column)  # not available due to raiseload=True
    print(a.id)

    a1 = session.execute(select(A).where(A.id == 1).execution_options(populate_existing=populate_existing)).scalar_one()

    # print(a1.other_column)  # not available due to raiseload=True
    print(a1.id)

    a2 = session.execute(select(A).where(A.id == 1).options(undefer(A.other_column)).execution_options(populate_existing=populate_existing)).scalar_one()

    print(a2.other_column)  # available now

    a3 = session.execute(select(A).where(A.id == 1).execution_options(populate_existing=populate_existing)).scalar_one()

    assert a3 is a2
    print(a2.other_column)  # <-- fails if populate_existing is True, works if False


run(False)
```

We're a little nervous about this other situation (which is not hypothetical / has been extracted from a real world test if we start using populate_existing in more places) where due to the combination of the identity map + populate_existing seeming to fully replace objects, we can run into situations with it where a function might call another function that queries something and _removes_ loaded columns out from under the first function. (We could remove the raiseloads and have this lead to just a lazy query, but we'd like to be very explicit as to when we're making network requests / when we aren't.)

The ideal behavior for us would be, I think, that either:
* these two things _did not interfere with eachother_ (e.g. `a3 is not a2`) -- but suspect that's more or less impossible due to how sqlalchemy works
* the `populate_existing` in `a2` _only_ overwrites columns that are actually returned (e.g. `update` semantics not `set` semantics)

Is either of those things possible, or are we thinking about this wrong?

**Reply by zzzeek** [reply]
populate existing means the current in-memory state you have is not what you want, you want the DB's state.  so if you want other_column to be there also you have to undefer it again.  if a2 and a3 are really in separate streams of use, and a2 has to keep being used, you could turn a2 into a detached object, or use two sessions perhaps.      at the very least, a reason to not be "nervous" about the above scenario is that it's deterministic and would show up in your testing ahead of time, it should not be random, unless the queries that are conflicting here are against randomly overlapping result sets.

overall if you really need to SELECT..FOR UPDATE a certain row, that's a pretty specific thing and usually I'd expect that transaction would not have a lot of unrelated things going on in it.

**Reply by ay-ramp** [reply]
unfortunately that is a bit optimistic for us, although we may be a bit of an edge case for you. we're operating in a million+ LOC monolith with hundreds of engineers working on it, where there is lots of code like:

```
def route_handler(object_id, param1, param2, param3, ...):
  # this whole route happens in one big transaction for atomicity purposes
  x = db.session.execute(select(Model).where(Model.id == object_id)).scalar_one()
  if param1:
      # this will call into code that's owned by a different team than the route_handler ower
     do_y(object_id, param1, param2)
  if param2:
      # this will call into code that's owned by a third different team than the route_handler ower
     do_z(object_id, param1, param2)
  ... # etc
 db.session.commit()
```

and we will often only take out select for update type locks in one of many possible branches of functions called by e.g. `do_y`, and we don't often have full integration-level coverage of every permutation of state that could happen / rely on more localized unit tests.

open to the opinion that we are not your core users and will need to figure something out on our own :) but would be great if there was something that reduced these edges for us / you pointed us in the right direction.

**Reply by ay-ramp** [reply]
is there any way for us to get all of the different load options that were applied on an existing instance, so that we can safely always requery them? e.g. if we `select(MyModel).options(undefer(MyModel.xyz))`, maybe we can look at somethng in internal state and subsequently always do something like `select(MyModel).options(*existing_options(MyModel, object_id))`?

**Reply by zzzeek** [reply]
I think the better way to do this is to isolate out the "with_for_update" / "populate_existing" case, if only do_z() needs that case, they should do that separately and they should likely name the columns they care about.  Here's a demonstration of three functions that each work on a raiseload attribute.  one function cares about FOR UPDATE for a specific attribute, so it names that attribute using `load_only()`, leaving the other attributes unaffected:

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import select
from sqlalchemy import update
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import load_only
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session
from sqlalchemy.orm import undefer


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    x: Mapped[int] = mapped_column(deferred_raiseload=True)
    y: Mapped[int] = mapped_column(deferred_raiseload=True)
    z: Mapped[int] = mapped_column(deferred_raiseload=True)


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

with Session(e) as sess:
    sess.add(A(x=1, y=2, z=3))
    sess.commit()


def do_x(sess: Session, a_obj: A):
    print(f"x is: {a_obj.x}")


def do_y(sess: Session, a_obj: A):
    # do out-of-trans update, to have the with_for_update be meaningful
    with e.begin() as conn:
        conn.execute(update(A).values(y=7))

    sess.execute(
        select(A)
        .with_for_update()
        .options(load_only(A.id, A.y))
        .execution_options(populate_existing=True)
    ).scalar_one()

    assert a_obj.y == 7
    print(f"y is: {a_obj.y}")


def do_z(sess: Session, a_obj: A):
    print(f"z is: {a_obj.z}")


with Session(e) as sess:
    a_obj = sess.execute(
        select(A).options(undefer(A.x), undefer(A.y), undefer(A.z))
    ).scalar_one()
    do_x(sess, a_obj)
    do_y(sess, a_obj)
    do_z(sess, a_obj)

    sess.commit()

```

**Reply by ay-ramp** [reply]
ahh, `load_only` is exactly what we were looking for here. looks very promising - thank you very much as always!

**Reply by CaselIT** [reply]
Oh interesting, I did not know that load_only worked that way.
I'll improve the docs for it

**Reply by CaselIT** [reply]
added a mention here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5232

**Reply by ay-ramp** [reply]
Okay, think that we are almost fully in a good place with this, but running into one last snag which is preserving _relationships_. It seems like even with the `load_only` in place, a relationship that is previously loaded onto the object will get nulled out if a subsequent call with `populate_existing()` happens that doesn't include the relationship as a load option. Since we're using `raiseload` on many of these, that can cause some unexpected errors to pop up.

Any advice?

(If this sounds like unexpected behavior, can isolate a repro from your example above.)

**Reply by zzzeek** [reply]
load_only() only affects column-based attributes.  for relationships, send along `lazyload('*')`.

**Reply by ay-ramp** [reply]
Got it, makes sense. Thank you! I can't do that directly, because if a relationship _isn't_ loaded I'd like to preserve the `lazy="raise"` behavior.

Here's a general util that I've written that seems to work:

```py
def _all_selectable_attributes(model_cls: Type[Base]) -> Iterable[InstrumentedAttribute]:
    for k, maybe_col in inspect(model_cls).mapper.columns.items():
        if not isinstance(maybe_col, Column):
            # Skip hybrid properties, column properties, relationships, etc
            continue

        # The instrumented attribute is the actual attribute on the model class, which
        # wraps the Column with other ORM info
        attr = getattr(model_cls, k)
        assert isinstance(attr, InstrumentedAttribute)

        if attr.property.deferred:
            # we only want to include properties that would normally be loaded, not
            # deferred ones
            continue

        yield attr


def _all_relationship_attrs(model_cls: Type[Base]) -> Iterable[InstrumentedAttribute]:
    for k, maybe_rel in inspect(model_cls).mapper.relationships.items():
        attr = getattr(model_cls, k)
        assert isinstance(attr, InstrumentedAttribute)

        yield attr

M = TypeVar("M", bound=Base)
def _loaded_relationship_attrs(model_cls: Type[M], unloaded_keys: set[str]) -> Iterable[InstrumentedAttribute]:
    for k, maybe_rel in inspect(model_cls).mapper.relationships.items():
        if k in unloaded_keys:
            # Skip unloaded relationships
            continue

        attr = getattr(model_cls, k)
        assert isinstance(attr, InstrumentedAttribute)

        yield attr

def _primary_key_clause(model_cls: Type[Base], ident: object) -> Any:
    pk_cols: tuple[Column[Any], ...] = inspect(model_cls).primary_key
    pk_attrs = [getattr(model_cls, col.name) for col in pk_cols]
    if len(pk_attrs) == 1:
        return pk_attrs[0] == ident
    else:
        return and_(
            *(attr == ident for attr in pk_attrs)
        )



def safe_populate_existing_select(model_cls: Type[Base], ident: object) -> Select:
    """Generate a select statement for a single instance of the model_cls with the primary key identified in ident"""
    key = identity_key(model_cls, ident)
    if (instance := session.identity_map.get(key)) is not None:
        unloaded_keys = inspect(instance).unloaded
        loaded_attrs = _loaded_relationship_attrs(model_cls, unloaded_keys)
    else:
        loaded_attrs = set()

    return select(model_cls).where(_primary_key_clause(model_cls, ident)).options(
        load_only(*(_all_selectable_attributes(model_cls))),
        *(lazyload(attr) for attr in loaded_attrs)
    ).execution_options(populate_existing=True)
```

although this is doing...a lot more work/inspection logic than I'd have hoped, it seems to reliably result in the right (minimal) sql being generated and instance state not getting clobbered between the two for columns, deferred columns, and relationships.

Anything that you'd caution me about on this approach? (Would also welcome any simpler versions.)

**Reply by CaselIT** [reply]
If I understand correct load_only is interfering with the state of a relationship that was not listed to on its attributes, meaning that for a class with `col1,col2,rel1` `load_only(col1)` would reset the state of `rel1` but not mess with `col2`. @ay-ramp can you confirm that this is the behaviour you are seeing?

@zzzeek if this is the case I think rel1 should be left alone here, since no one told it to change.

**Reply by ay-ramp** [reply]
>If I understand correct load_only is interfering with the state of a relationship that was not listed to on its attributes, meaning that for a class with col1,col2,rel1 load_only(col1) would reset the state of rel1 but not mess with col2. @ay-ramp can you confirm that this is the behaviour you are seeing?

yes, that is correct -- which is solvable it seems like through the `*(lazyload(attr) for attr in loaded_attrs` in the snipped above

**Reply by CaselIT** [reply]
yes, but I would argue that it should not be needed.

I kinda understand why it does that, `load_only` influences only the columns, ignoring all relationships that would continue to have the default populate existing behaviour.
But I would argue that this behaviour is unhelpful in the general case

**Reply by zzzeek** [reply]
That's a lot of code which I'd be uncomfortable with, but it also doesnt seem to be doing what we talked about.     I dont see where you are doing the actual for_update() part, and also I dont see where the caller would be indicating the specific attributes they care about for the "for_update", the idea was that someone using select.for_update() should *know* what specific columns they care about for them to be using SELECT..FOR UPDATE.   if you simply want already-populated columns to remain untouched, then just dont use populate_existing().   The whole point of using populate_existing() with FOR UPDATE is the use case at #5572, where there's specific columns they are looking for to have a specific state that is guaranteed in that transaction, independent of existing data that might have been loaded there from a lower isolation level.   

also from the code above it looks like you are operating on only a single primary key identity at a time.   if your use case involves just one row in all cases, that gives you more options since you can pre-expire the attributes you care about and skip the populate_existing part.     you can just pull it from the identity map refresh those attributes:

```py
from sqlalchemy.orm.util import identity_key

def do_y(sess: Session, ident: int):

    a_obj = sess.identity_map.get(identity_key(A, ident), None)
    if a_obj is None:
        a_obj = sess.get_one(A, ident, with_for_update=True)
    else:
        sess.refresh(a_obj, attribute_names=["id", "y"], with_for_update=True)

```

I think what's important here is setting up what is the input for what we are doing here:

1. are we doing FOR UPDATE on only one row at a time?
2. do we already have the object for that row or do we just have a PK identity?
3. why are we doing FOR UPDATE and what attributes do we need to be refreshed with the FOR UPDATE row?

**Reply by ay-ramp** [reply]
> the idea was that someone using select.for_update() should know what specific columns they care about for them to be using SELECT..FOR UPDATE

this might be nice in theory (and we could try to guarantee this), but the behavior we'd really like for mental model simplicity is- "all the data that comes back over the wire from the database gets written to the object, but leave everything else alone"). if we _don't_ have to specify attributes ahead of time, that'd be a win (which is what the util is trying to do)

--

re: 1 and 2, yeah! in our use case at least, we only want to fetch one row at a time; sometimes we'll have the object but most of the time we'll just have a PK identity. your implementation looks a lot more promising than mine with that in mind. doing some more testing...

**Reply by zzzeek** [reply]
I can certainly see in the general sense how populate_existing erasing the loaded relationships is something people might not want to have, although that's more about collection-based attributes.  for a many-to-one relationship, if you are looking for columns to be refreshed then you'd want many-to-ones refreshed also.

I was sort of expecting `noload('*')` to instruct that the relationships should be left alone but it doesnt do that, it actually blanks them out.   

I think we should think about making some nicer loader options that can supersede populate_existing altogether with more specific instructions.   I think this option would meet all of these needs:

 `stmt.options(load_only('a', 'b', 'c', populate_existing=True))`   - a load only plus populate existing for those named attributes

**Reply by ay-ramp** [reply]
hmm, `session.refresh` _without_ named attributes still clobber previously loaded relationships, so i think if we didn't want to force people to manually try to define what columns they care about synchronizing, then we'd still need to be doing the inspection stuff ala the snippet above?

**Reply by ay-ramp** [reply]
FWIW- we don't have to use `load_only()` anywhere else, and while it's great that we can use it as an escape hatch, it would be less surprising at least to this user to have some kind of flag that's like `update_existing=True` that directly gives the "patch existing" vs "replace existing" semantics that we're really trying to get.

**Reply by zzzeek** [reply]
I'm still at the beginning here.  why do you need SELECT...FOR UPDATE if the user does not know what columns need to be synchronized?

**Reply by ay-ramp** [reply]
>why do you need SELECT...FOR UPDATE if the user does not know what columns need to be synchronized?

we don't, in the same way that technically we don't necessarily need to load every column from sqlalchemy before selecting it and using it in code. but not having to specify that information upfront is helpful, and we're dealing with lots of existing code that does things like:

```
def do_y(obj_id):
  a = db.session.execute(select(A).where(A.id==obj_id).with_for_update()).scalar_one()
  # call lots of subfunctions (maybe many layers) that could use any number of fields on A and 
  # have been written over years to assume that a _is_ the latest locked state from the DB
```

which works if `a` exists in the identity map; and doesn't if it doesn't. 

the ideal state would be that we don't have to think about whether it was fetched previously or not; and that subsequent callers (outside of `do_y`) don't need to think about what `do_y()` does / doesn't do either.

**Reply by zzzeek** [reply]
im not clear on why 

```
a = db.session.execute(select(A).where(A.id==obj_id).with_for_update()).scalar_one()
```

doesnt work if obj_id is not in the identity map.

I've scrolled up top and it seems the problem in question is, "we found a particular FOR UPDATE query that was not effective because we didnt use populate existing.  so now we'd like some way to prevent this same issue from happening elsewhere."    What's not totally clear is the scope at which you are willing to *change* code.    It seems like you have queries that are of the form `select(ent).where(<primary key of a single instance>)` which are the kinds of queries you want to target.      and that you are willing to change that particular call.    so I proposed using refresh() instead.   "session.refresh without named attributes still clobber previously loaded relationships" - I proposed sending the named attributes.  if you want all loaded columns, you can get that like this:

```
        attr_names = set(inspect(a_obj).mapper.column_attrs.keys()).difference(inspect(a_obj).unloaded)
        sess.refresh(a_obj, attribute_names=attr_names, with_for_update=True)

```


What I am not comfortable with at the moment is rushing out a new API that suits some specific combination of uses here which will again put us in the position of having *more* APIs that are *still not enough*.  The fact that we can already see there are so many ways to get at the behavior we want here, all of them falling short, is a product of this kind of feature adds - that is, one offs to suit someone's particular immediate need.

**Reply by ay-ramp** [reply]
> im not clear on why

> a = db.session.execute(select(A).where(A.id==obj_id).with_for_update()).scalar_one()
> doesnt work if obj_id is not in the identity map.

sorry, dashed that last one off too quickly -- meant the opposite (as you deduced later). apologies for the lack of clarity.

> What I am not comfortable with at the moment is rushing out a new API that suits some specific combination of uses here which will again put us in the position of having more APIs that are still not enough. The fact that we can already see there are so many ways to get at the behavior we want here, all of them falling short, is a product of this kind of feature adds - that is, one offs to suit someone's particular immediate need.

agreed, and also apologize if I'm coming across too pushily here :) know changing APIs is a big deal, the suggestions above were meant in a longer-term spirit and as an alternative to other API changes being suggested that relied more heavily on `load_only`.

----

```
M = TypeVar("M", bound=Base)
def safe_get_for_update(model_cls: Type[M], ident: object) -> M:
    maybe_obj: M | None = session.identity_map.get(identity_key(model_cls, ident), None)
    if maybe_obj is None:
        # The object hasn't been fetched yet, so we can return directly
        return session.get(model_cls, ident, with_for_update=True)

    # Refetch all of the columns that have been loaded so far, and lock the row
    attr_names = set(inspect(model_cls).mapper.column_attrs.keys()).difference(inspect(model_cls).unloaded)
    session.refresh(maybe_obj, attribute_names=attr_names, with_for_update=True)
    return maybe_obj
```

this seems like a good general purpose util that we can start using places and unblocks us. just ran it against the various test cases we've been worrying about and seems to be quite robust (and much simpler than what we would have have had to do). appreciate the guidance and discussion as always!

**Reply by zzzeek** [reply]
@CaselIT I think we can perhaps make use of either typeddicts or literals and allow populate_existing parameter as a set or dict with whatever options we need

```
select(..).execution_options(populate_existing={"columns": "all_existing"})    # overwrite already loaded columns, no relationships
select(..).execution_options(populate_existing={"columns": "unmodified", "relationships": "existing_manytoone"})    # overwrite unmodified columns, already-loaded many-to-one relationships
select(..).execution_options(populate_existing={"relationships": "all_existing"}) # overwrite already-loaded relationships
```

but this is definitely 2.1 stuff.  not expected for many months

**Comment by CaselIT** [comment]
Let's continue here from https://github.com/sqlalchemy/sqlalchemy/discussions/11162#discussioncomment-8867906

>

> @CaselIT I think we can perhaps make use of either typeddicts or literals and allow populate_existing parameter as a set or dict with whatever options we need
> 
> ```
> select(..).execution_options(populate_existing={"columns": "all_existing"})    # overwrite already loaded columns, no relationships
> select(..).execution_options(populate_existing={"columns": "unmodified", "relationships": "existing_manytoone"})    # overwrite unmodified columns, already-loaded many-to-one relationships
> select(..).execution_options(populate_existing={"relationships": "all_existing"}) # overwrite already-loaded relationships
> ```
> 
> but this is definitely 2.1 stuff. not expected for many months

I think that would work and is likely quite nice to use with typing using IDE etc.

Use cases may be as follows:
Columns:
- loaded: overwrite *only* already loaded columns
- all: overwrite all columns, even if unloaded
- unmodified: overwrite all unmodified columns, even if unloaded (only useful with autoflush=False)
- unmodified_loaded: overwrite *only* unmodified already loaded columns (only useful with autoflush=False)
- none: don't reload any column. This could be the default when using a dict

Relationship
- loaded: overwrite *only* already loaded relationship
- loaded_manytoone: overwrite *only* already loaded relationship many to one
- unmodified_loaded: overwrite *only* unmodified already loaded relationship (only useful with autoflush=False ?)
- none: don't reload any relationship. This could be the default when using a dict

I'm sure I've missed some cases above.
Covering the complete use cases with an api seems quite complex to do with only string constant. It's even more complex but more expressive to have nested dict too. For example:
```ts
interface PopulateExisting {
  populate_existing: boolean | {
    column?: 'none' | {
      kind: 'all' | 'loaded';
      ignore_dirty: boolean;
    };
    relationship?: 'none' | {
      kind: 'all' | 'loaded';
      ignore_dirty: boolean; // not sure if it makes sense here
      direction: 'all' | 'many-to-one' | 'one-to-many' | 'many-to-many';
    }
  }
}
```
